### PR TITLE
[NewOptimizer] Various misc bugfixes

### PR DESCRIPTION
**Pull request by Keno**
Some fixes @vtjnash and I worked on last week that accidentally got dropped from the last roundup commit, as well as one new commit to address a failure I saw on kf/iterate.

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -6610,10 +6611,29 @@ static std::unique_ptr<Module> emit_function(
...
+                        // so we need to make sure that such values are legal to be copied from.
+                        if (VN->getType() == T_prjlvalue) {
+                            VNUndef =  (llvm::Value*)ConstantPointerNull::get(cast<PointerType>(T_prjlvalue));
+                        } else if (VN->getType()->isPointerTy()) {
+                            PointerType *PTy = cast<PointerType>(VN->getType());
+                            VNUndef = new AllocaInst(PTy->getElementType(),
+#if JL_LLVM_VERSION >= 50000
+                            0,
+#endif
+                            "undef_alloca", /*InsertBefore=*/ctx.ptlsStates);
```
Comment (vtjnash) [inline src/codegen.cpp]: indent

File: src/codegen.cpp [hunk 2]
```diff
@@ -6610,10 +6611,29 @@ static std::unique_ptr<Module> emit_function(
...
+                    if (!VNUndef) {
+                        // While we're guaranteed that this value will not be looked at dynamically,
+                        // we're not guaranteed that it won't be moved around. For things allocated on the
+                        // stack, that generally involves memcpying them from one stack slot to another,
+                        // so we need to make sure that such values are legal to be copied from.
+                        if (VN->getType() == T_prjlvalue) {
+                            VNUndef =  (llvm::Value*)ConstantPointerNull::get(cast<PointerType>(T_prjlvalue));
+                        } else if (VN->getType()->isPointerTy()) {
+                            PointerType *PTy = cast<PointerType>(VN->getType());
+                            VNUndef = new AllocaInst(PTy->getElementType(),
```
Comment (vtjnash) [inline src/codegen.cpp]: I don't think we do / can correctly set the element type in many cases (especially with union-alloca)

Comment (Keno) [reply src/codegen.cpp]: Ah, fair point, we can just get the size from the union type. Alternatively, we could just emit one alloca big enough for all phis at the top.

**Comment by Keno** [comment]
Fixed to use one big undef alloca with appropriate size instead.

**Comment by vtjnash** [comment]
Won’t that cause issues for SROA?

**Comment by Keno** [comment]
I put a lifetime.end intrinsic at the top of the function, so it's legal for SROA to ignore any loads/stores known to be to that alloca.

**Comment by Keno** [comment]
This has a good number of CI failures. The only one I'm not familiar with is a missing gc root in ccall, which @vtjnash says is fixed by #26732. Will merge once @vtjnash has had a chance to review.

**Review by vtjnash — APPROVED** [review]

**Pull request merged**